src/Device/Driver/FANET/Parser.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2016 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "Device.hpp"
+#include "NMEA/Info.hpp"
+#include "Parser.hpp"
+#include "FANET/FanetAddress.hpp"
+
+/*
+ * Updates existing station or allocates new one
+ * Returns false if the list is full, true otherwise
+ */
+bool
+FanetDevice::AddStationToList(StationsList &stations_list, double clock,
+                              const FanetStation &station) {
+  FanetStation *fanet_slot = stations_list.FindStations(station.address);
+
+  if (fanet_slot == nullptr) {
+    fanet_slot = stations_list.AllocateStation();
+    if (fanet_slot == nullptr)
+      // no more slots available
+      return false;
+
+    fanet_slot->Clear();
+    fanet_slot->address = station.address;
+    fanet_slot->location = station.location;
+    fanet_slot->location_available = station.location_available;
+  }
+
+  fanet_slot->valid.Update(clock);
+  fanet_slot->Update(station);
+
+  return true;
+}
+
+bool
+FanetDevice::IntegrityCheck(char *cmd_str) {
```

Comment:
This method returns `true` given an empty string. Is that desired?

Reply:
No, fixed in recent amend

Reply:
Using `strlen()==0` is a pretty clumsy way to check for empty strings. If it's not empty, it needs to check each and every character in the string. The best way to check for empty strings in C is `*cmd_str == 0`.
But you don't even neeed that. You can just change the loop, e.g.:
```
do {
  if (!IsHexOrCommaChar(*cmd_str))
    return false;
} while (*++cmd_str != 0);
return true;
```
If the first character is a null byte, it will fail the test. But follow-up null bytes will instead just break out of the loop.
My example also avoids your quirky `continue` with `return` at the end of the loop scope... which I think is bad coding style.